src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastCache.cs

```diff
@@ -0,0 +1,461 @@
+
+#if CORECLR
+                // in CoreCLR we do ordinary reads of the entry parts and
+                // Interlocked.ReadMemoryBarrier() before reading the version
+                nuint entrySource = pEntry._source;
+#else
+                // must read this before reading the version again
+                nuint entrySource = Volatile.Read(ref pEntry._source);
+#endif
+
+                // mask the lower version bit to make it even.
+                // This way we can check if version is odd or changing in just one compare.
+                version &= unchecked((uint)~1);
+
+                if (entrySource == source)
+                {
+
+#if CORECLR
+                    // in CoreCLR we do ordinary reads of the entry parts and
+                    // Interlocked.ReadMemoryBarrier() before reading the version
+                    nuint entryTargetAndResult = pEntry._targetAndResult;
+#else
+                    // must read this before reading the version again
+                    nuint entryTargetAndResult = Volatile.Read(ref pEntry._targetAndResult);
+#endif
+
+                    // target never has its lower bit set.
+                    // a matching entryTargetAndResult would the have same bits, except for the lowest one, which is the result.
+                    entryTargetAndResult ^= target;
+                    if (entryTargetAndResult <= 1)
+                    {
+                        // make sure the second read of 'version' happens after reading 'source' and 'targetAndResults'
+                        //
+                        // We can either:
+                        // - use acquires for both _source and _targetAndResults or
+                        // - issue a load barrier before reading _version
+                        // benchmarks on available hardware (Jan 2020) show that use of a read barrier is cheaper.
+
+#if CORECLR
+                        Interlocked.ReadMemoryBarrier();
+#endif
+
+                        if (version != pEntry._version)
+                        {
+                            // oh, so close, the entry is in inconsistent state.
+                            // it is either changing or has changed while we were reading.
+                            // treat it as a miss.
+                            break;
+                        }
+
+                        return (CastResult)entryTargetAndResult;
+                    }
+                }
+
+                if (version == 0)
+                {
+                    // the rest of the bucket is unclaimed, no point to search further
+                    break;
+                }
+
+                // quadratic reprobe
+                i++;
+                index = (index + i) & TableMask(ref tableData);
+            }
+            return CastResult.MaybeCast;
+        }
+
+        // the rest is the support for updating the cache.
+        // in CoreClr the cache is only updated in the native code
```

Comment:
Would it make sense for CoreCLR to just call this managed implementation instead?

Reply:
Interesting idea. I think what could prevent that is: 

- if we need to use the cache before we can run managed code. 
JIT calls into casting machinery, so it is hard to tell if we may need the cache too early. 

- if we need to use the cache before managed cctor has run. 
I'd guess the cctor could be forced to run early enough.

- if we need to use cache from some mode that does not allow managed code.
Most likely this is not a requirement, but I am not sure.

- if calling managed code from native has high overhead.
Adding to the cache is reasonably fast. We do not see the cost of adding stuff even at start up, when we'd expect cache misses.  
I do not have a good sense of how expensive it is to call managed code from the native runtime.

Would we want to call managed TryGet as well?  
(that would be a lot more sensitive to the overhead and TryGet is `GC_NOTRIGGER`, I vaguely remember there were some reasons for that.)

Reply:
I think this is worth considering, but it should be a separate change. 

It does not look like a straightforward tweak and may need some experimenting. 
Also the impact of such change will be mostly in CoreClr, while this PR mostly affects NativeAOT, so in terms of watching for failures or stress/perf consequences, it would be better to have separate changes.

Reply:
Logged an issue to follow up on this - https://github.com/dotnet/runtime/issues/84448